compiler/semcall.nim

```diff
@@ -247,7 +247,13 @@ proc presentFailedCandidates(c: PContext, n: PNode, errors: CandidateErrors):
           candidates.add typeToString(got)
           candidates.addDeclaredLocMaybe(c.config, got)
           doAssert wanted != nil
-          if got != nil: effectProblem(wanted, got, candidates, c)
+          if got != nil:
+            if got.kind == tyProc and wanted.kind == tyProc:
+              # These are proc mismatches so,
+              # add the extra explict detail of the mismatch
+              candidates.addPragmaAndCallConvMismatch(wanted, got, c.config)
+            effectProblem(wanted, got, candidates, c)
+
```

Comment:
can be handled in future work, but should we also handle tyIterator? eg:
```nim
iterator bar1(): int {.closure.} = discard
var s = ["ab", "cd"]
iterator bar2(): int {.closure.} =
  echo s
proc call(a: iterator(): int {.nosideeffect, gcsafe.}) = discard
call(bar1)
call(bar2)
```
currently gives:
```
Error: type mismatch: got <iterator (): int{.closure, locks: 0.}>
but expected one of:
proc call(a: iterator (): int {.nosideeffect, gcsafe.}) [proc declared in /Users/timothee/git_clone/nim/timn/tests/nim/all/t12520.nim(9, 6)]
  first type mismatch at position: 1
  required type for a: iterator (): int{.closure, noSideEffect, gcsafe.} [proc]
  but expression 'bar2' is of type: iterator (): int{.closure, locks: 0.} [proc]
  This expression is not GC-safe. Annotate the proc with {.gcsafe.} to get extended error information.
```
which, incidentally, is not bad but doesn't seem to use your PR's code

Reply:
Due to the fact that `tyProc` also counts iterators teh above code outputs:
```
Error: type mismatch: got <iterator (): int{.closure, locks: 0.}>
but expected one of: 
proc call(a: iterator (): int {.nosideeffect, gcsafe.})
  first type mismatch at position: 1
  required type for a: iterator (): int{.closure, noSideEffect, gcsafe.}
  but expression 'bar2' is of type: iterator (): int{.closure, locks: 0.}
  Pragma mismatch: got '{..}', but expected '{.noSideEffect, gcsafe.}'.
  This expression is not GC-safe. Annotate the proc with {.gcsafe.} to get extended error information.
  ```

Reply:
hmm, IIRC it didn't show `Pragma mismatch: got '{..}', but expected '{.noSideEffect, gcsafe.}'.` when i had checked, but yes, looks like it works now, so, ya, done!